plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenGroupFacet.java

```diff
@@ -0,0 +1,202 @@
+                                       final LinkedHashMap<Repository, Response> responses) throws IOException
+  {
+    checkArgument(mavenFacet.getMavenPathParser().isRepositoryMetadata(mavenPath),
+        "Only metadata can be merged and cached: %s", mavenPath);
+    checkArgument(!mavenPath.isSubordinate(), "Only metadata XML can be merged and cached: %s", mavenPath);
+    final LinkedHashMap<Repository, Content> metadataContents = Maps.newLinkedHashMap();
+    for (Map.Entry<Repository, Response> entry : responses.entrySet()) {
+      if (entry.getValue().getStatus().getCode() == HttpStatus.OK) {
+        final Response response = entry.getValue();
+        if (response instanceof PayloadResponse && ((PayloadResponse) response).getPayload() instanceof Content) {
+          metadataContents.put(entry.getKey(), (Content) ((PayloadResponse) response).getPayload());
+        }
+      }
+    }
+
+    if (metadataContents.isEmpty()) {
+      return null;
+    }
+    return doMergeAndCacheMetadata(mavenPath, metadataContents);
+  }
+
+  private Content doMergeAndCacheMetadata(final MavenPath mavenPath,
+                                          final LinkedHashMap<Repository, Content> metadataContents) throws IOException
+  {
+    if (metadataContents.size() == 1) {
+      return metadataContents.get(0);
+    }
+    final MetadataXpp3Reader reader = new MetadataXpp3Reader();
+    List<MetadataEnvelope> metadatas = Lists.newArrayList();
+    for (Map.Entry<Repository, Content> entry : metadataContents.entrySet()) {
+      final String origin = entry.getKey().getName() + " @ " + mavenPath.getPath();
+      try (InputStream inputStream = entry.getValue().openInputStream()) {
+        final Metadata metadata = reader.read(inputStream);
+        metadatas.add(new MetadataEnvelope(origin, metadata));
+      }
+      catch (XmlPullParserException e) {
+        // skip it, log it
+        log.info("Unparseable repository metadata: {}", origin, e);
+      }
+    }
+
+    final Metadata mergedMetadata = metadataMerger.merge(metadatas.toArray(new MetadataEnvelope[metadatas.size()]));
+    if (mergedMetadata == null) {
+      return null;
+    }
+    final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    new MetadataXpp3Writer().write(byteArrayOutputStream, mergedMetadata);
+    // Metadata is rather small, so let's do it like this
+    final Iterable<HashAlgorithm> hashAlgorithms = Iterables.transform(Arrays.asList(HashType.values()),
+        new Function<HashType, HashAlgorithm>()
+        {
+          @Override
+          public HashAlgorithm apply(final HashType input) {
+            return input.getHashAlgorithm();
+          }
+        });
+    final Map<HashAlgorithm, HashCode> hashCodes = Hashes.hash(hashAlgorithms,
+        new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+    final Content content = new Content(
+        new BytesPayload(
+            byteArrayOutputStream.toByteArray(),
```

Comment:
minor, but no need to create the byte array twice, assign to var should save us some memory